crates/tx5/src/backend.rs

```diff
@@ -0,0 +1,134 @@
+            #[cfg(feature = "backend-go-pion")]
+            Self::GoPion => go_pion::default_config(),
+            #[cfg(feature = "backend-webrtc-rs")]
+            Self::WebrtcRs => todo!(),
+            Self::Mock => serde_json::json!({}),
+        }
+    }
+
+    /// Connect a new backend module endpoint.
+    pub async fn connect(
+        &self,
+        url: &str,
+        listener: bool,
+        config: &Arc<Config>,
+    ) -> Result<(DynBackEp, DynBackEpRecv)> {
+        match self {
+            #[cfg(feature = "backend-go-pion")]
+            Self::GoPion => go_pion::connect(config, url, listener).await,
+            #[cfg(feature = "backend-webrtc-rs")]
+            Self::WebrtcRs => todo!(),
+            Self::Mock => todo!(),
+        }
+    }
+}
+
+/// Backend connection.
+pub trait BackCon: 'static + Send + Sync {
+    /// Send data over this backend connection.
+    fn send(&self, data: Vec<u8>) -> BoxFuture<'_, Result<()>>;
+
+    /// Get the pub_key identifying this connection.
+    fn pub_key(&self) -> &PubKey;
+
+    /// Returns `true` if we successfully connected over webrtc.
+    // TODO - this isn't good encapsulation
+    fn is_using_webrtc(&self) -> bool;
+
+    /// Get connection statistics.
+    // TODO - this isn't good encapsulation
+    fn get_stats(&self) -> tx5_connection::ConnStats;
+}
+
+/// Trait-object version of backend connection.
+pub type DynBackCon = Arc<dyn BackCon + 'static + Send + Sync>;
+
+/// Backend connection receiver.
+pub trait BackConRecv: 'static + Send {
+    /// Receive data from this backend connection.
+    fn recv(&mut self) -> BoxFuture<'_, Option<Vec<u8>>>;
+}
+
+/// Trait-object version of backend connection receiver.
+pub type DynBackConRecv = Box<dyn BackConRecv + 'static + Send>;
+
+/// Pending connection.
+pub trait BackWaitCon: 'static + Send {
+    /// Wait for the connection
+    fn wait(
+        &mut self,
+        // TODO - this isn't good encapsulation
+        recv_limit: Arc<tokio::sync::Semaphore>,
+    ) -> BoxFuture<'static, Result<(DynBackCon, DynBackConRecv)>>;
+
+    /// Get the pub_key identifying this connection.
+    fn pub_key(&self) -> &PubKey;
+}
+
+/// Trait-object version of backend wait con.
+pub type DynBackWaitCon = Box<dyn BackWaitCon + 'static + Send>;
+
+/// Backend endpoint.
+pub trait BackEp: 'static + Send + Sync {
+    /// Establish an outgoing connection from this backend endpoint.
+    fn connect(&self, pub_key: PubKey)
+        -> BoxFuture<'_, Result<DynBackWaitCon>>;
+
+    /// Get the pub_key identifying this endpoint.
+    fn pub_key(&self) -> &PubKey;
+}
+
+/// Trait-object version of backend endpoint.
+pub type DynBackEp = Arc<dyn BackEp + 'static + Send + Sync>;
+
+/// Backend endpoint receiver.
+pub trait BackEpRecv: 'static + Send {
+    /// Receive incoming connection from this backend endpoint.
+    fn recv(&mut self) -> BoxFuture<'_, Option<DynBackWaitCon>>;
+}
+
+/// Trait-object version of backend endpoint receiver.
```

Comment:
What's the difference between this and the backend connection receiver?

Reply:
The endpoint receiver receives incoming connections. The connection receiver receives incoming messages over that one connection.

Reply:
Now that I look again at the trait functions it's pretty clear. I guess the trait names could speak more to that, e. g. `ConnRecv` and `DataRecv`

Reply:
I'd personally find that confusing, since then you'd be given a `ConnRecv` when creating an endpoint, and not when opening a connection...

Reply:
Right, maybe `EpConnRecv` would be better. I find more confusing that a backend connection receiver does not receive connections, but messages/data. =D

Reply:
Fair enough: https://github.com/holochain/tx5/pull/105/commits/885af263ffc4b31fb27e37c9a67d74f983df79c1

Went with:
- `EpRecvCon`
- `ConRecvData`